In [4]:
"""
    dataset information:
        train: 29837
        dev: 4475
        test: 2985
"""

import pandas as pd
import json


# training data

with open('dataset/MathQA/train.json') as f:
    data = json.loads(f.read())
    
train_df = pd.DataFrame(data)

print('training problems number: ', train_df['Problem'].count())
print(train_df.groupby(['category'])['Problem'].count())

train_df.head(5)

training problems number:  29837
category
gain            5120
general        13273
geometry        2117
other           1814
physics         7063
probability      450
Name: Problem, dtype: int64


Problem  \
0  the banker ' s gain of a certain sum due 3 yea...   
1  average age of students of an adult school is ...   
2  sophia finished 2 / 3 of a book . she calculat...   
3                        120 is what percent of 50 ?   
4  there are 10 girls and 20 boys in a classroom ...   

                                           Rationale  \
0  "explanation : t = 3 years r = 10 % td = ( bg ...   
1  "explanation : let the original no . of studen...   
2  let xx be the total number of pages in the boo...   
3  "50 * x = 120 - - > x = 2.4 - - > 2.4 expresse...   
4  if girls is 10 and boys is 20 , then 10 / 20 ....   

                                             options correct  \
0  a ) rs . 400 , b ) rs . 300 , c ) rs . 500 , d...       a   
1  a ) 1200 , b ) 120 , c ) 360 , d ) 240 , e ) n...       d   
2    a ) 229 , b ) 270 , c ) 877 , d ) 266 , e ) 281       b   
3  a ) 5 % , b ) 240 % , c ) 50 % , d ) 2 % , e )...       b   
4  a ) 1 / 2 , b ) 1 / 3 , c ) 1 / 5 , d ) 10 / 3...       a   

                                   annotated_formula  \
0  divide(multiply(const_100, divide(multiply(36,...   
1  multiply(divide(subtract(multiply(add(32, 4), ...   
2        divide(90, subtract(const_1, divide(2, 3)))   
3               multiply(divide(120, 50), const_100)   
4                                     divide(10, 20)   

                                      linear_formula category  
0  multiply(n2,const_100)|multiply(n0,n1)|divide(...     gain  
1  add(n2,n3)|multiply(n1,n2)|multiply(n1,#0)|sub...  general  
2   divide(n0,n1)|subtract(const_1,#0)|divide(n2,#1)  general  
3              divide(n0,n1)|multiply(#0,const_100)|     gain  
4                                      divide(n0,n1)    other

In [5]:
# dev data

with open('dataset/MathQA/dev.json') as f:
    data = json.loads(f.read())
    
dev_df = pd.DataFrame(data)

print('dev problems number: ', dev_df['Problem'].count())
print(dev_df.groupby(['category'])['Problem'].count())

dev_df.head(5)

dev problems number:  4475
category
gain            743
general        2006
geometry        308
other           288
physics        1056
probability      74
Name: Problem, dtype: int64


Problem  \
0  a multiple choice test consists of 4 questions...   
1  a 3 - digit positive integer is chosen at rand...   
2  if x and y are positive integers and 7 + x + y...   
3  the hcf and lcm of two numbers m and n are res...   
4  in a kilometer race , a beats b by 48 meters o...   

                                           Rationale  \
0  "5 choices for each of the 4 questions , thus ...   
1  p ( three odd digits ) = 5 / 9 * 5 / 10 * 5 / ...   
2  "try each answer choices . for a : 7 + 3 + xy ...   
3  answer we have , m x n = 6 x 210 = 1260 ∴ 1 / ...   
4  "time taken by b run 1000 meters = ( 1000 * 12...   

                                             options correct  \
0    a ) 24 , b ) 120 , c ) 625 , d ) 720 , e ) 1024       c   
1  a ) 1 / 2 , b ) 31 / 36 , c ) 49 / 54 , d ) 7 ...       b   
2              a ) 3 , b ) 5 , c ) 6 , d ) 8 , e ) 9       c   
3  a ) 1 / 35 , b ) 3 / 35 , c ) 5 / 37 , d ) 2 /...       d   
4  a ) 238 sec , b ) 190 sec , c ) 667 sec , d ) ...       a   

                                   annotated_formula  \
0                                        power(5, 4)   
1  subtract(const_1, power(divide(const_1, const_...   
2                                   add(21, const_1)   
3                       divide(72, multiply(6, 210))   
4  subtract(divide(multiply(const_1, const_1000),...   

                                      linear_formula category  
0                                      power(n1,n0)|  general  
1  divide(const_1,const_2)|power(#0,n0)|subtract(...  general  
2                                   add(n1,const_1)|  general  
3                      multiply(n0,n1)|divide(n2,#0)  general  
4  divide(n0,n1)|multiply(const_1,const_1000)|div...  physics

In [6]:
# test data

with open('dataset/MathQA/test.json') as f:
    data = json.loads(f.read())
    
test_df = pd.DataFrame(data)

print('test problems number: ', test_df['Problem'].count())
print(test_df.groupby(['category'])['Problem'].count())

test_df.head(5)

test problems number:  2985
category
gain            558
general        1301
geometry        200
other           172
physics         702
probability      52
Name: Problem, dtype: int64


Problem  \
0  a shopkeeper sold an article offering a discou...   
1  what will be the difference between simple and...   
2  there are 28 stations between hyderabad and ba...   
3  the present population of a town is 3888 . pop...   
4                 the triplicate ratio of 1 : 9 is ?   

                                           Rationale  \
0  "giving no discount to customer implies sellin...   
1  "s . i . = ( 1000 * 14 * 4 ) / 100 = rs . 560 ...   
2  "the total number of stations = 30 from 30 sta...   
3  "p = 3888 r = 20 % required population of town...   
4                     "13 : 93 = 1 : 729 answer : e"   

                                             options correct  \
0  a ) 38 , b ) 27.675 , c ) 30 , d ) data inadeq...       a   
1    a ) 129 , b ) 130 , c ) 124 , d ) 133 , e ) 145       a   
2    a ) 156 , b ) 167 , c ) 870 , d ) 352 , e ) 380       c   
3  a ) 2500 , b ) 2100 , c ) 3500 , d ) 3600 , e ...       e   
4  a ) 1 : 0 , b ) 1 : 8 , c ) 1 : 7 , d ) 1 : 2 ...       e   

                                   annotated_formula  \
0  subtract(divide(multiply(add(const_100, 31.1),...   
1  subtract(subtract(multiply(1000, power(add(div...   
2  multiply(add(28, const_1), add(add(28, const_1...   
3   add(3888, divide(multiply(3888, 20), const_100))   
4   divide(power(const_2.0, 9), power(const_3.0, 9))   

                                      linear_formula category  
0  add(n1,const_100)|subtract(const_100,n0)|multi...     gain  
1  divide(n0,const_100)|add(#0,const_1)|multiply(...     gain  
2   add(n0,const_1)|add(#0,const_1)|multiply(#0,#1)|  physics  
3   multiply(n0,n1)|divide(#0,const_100)|add(n0,#1)|     gain  
4  power(const_2.0,n1)|power(const_3.0,n1)|divide...    other

### training data information:

##### count: 29837

    gain          5120
    general       13273
    geometry      2117
    other         1814
    physics       7063
    probability   450

In [13]:
import spacy

import torch
import torch.nn as nn
import torch.optim as optim

from torchtext.legacy.data import Field

In [14]:
spacy_en = spacy.load('en_core_web_sm')

In [15]:
def tokenize_en(text):
    return [tok.text for tok in spacy_en.tokenizer(text)]

SRC = Field(tokenize = tokenize_en, 
            init_token = '<sos>', 
            eos_token = '<eos>', 
            lower = True)

TGT = Field(tokenize = tokenize_en, 
            init_token = '<sos>', 
            eos_token = '<eos>', 
            lower = True)

In [20]:
train_src = train_df['Problem'].apply(tokenize_en)
print(train_src)

train_trg = train_df['Rationale'].apply(tokenize_en)
print(train_trg)

0        [the, banker, ', s, gain, of, a, certain, sum,...
1        [average, age, of, students, of, an, adult, sc...
2        [sophia, finished, 2, /, 3, of, a, book, ., sh...
3                      [120, is, what, percent, of, 50, ?]
4        [there, are, 10, girls, and, 20, boys, in, a, ...
                               ...                        
29832    [a, man, invests, some, money, partly, in, 10,...
29833    [average, of, 10, matches, is, 32, ,, how, man...
29834       [solve, below, question, 2, x, +, 1, =, -, 23]
29835                     [calculate, 85184, ÷, ?, =, 352]
29836    [7.51, 8.22, 7.86, 8.36, 8.09, 7.83, 8.30, 8.0...
Name: Problem, Length: 29837, dtype: object
0        [", explanation, :, t, =, 3, years, r, =, 10, ...
1        [", explanation, :, let, the, original, no, .,...
2        [let, xx, be, the, total, number, of, pages, i...
3        [", 50, *, x, =, 120, -, -, >, x, =, 2.4, -, -...
4        [if, girls, is, 10, and, boys, is, 20, ,, then...
            